src/cheri_cap_common.sail

```diff
@@ -61,30 +145,85 @@ function getCapHardPerms(cap) : Capability -> bits(cap_hperms_width) =
   @ bool_to_bits(cap.permit_execute)
   @ bool_to_bits(cap.global))
 
-/* Convert from capabilty struct to bits (no tag) */
-function capToBits(cap) : Capability -> CapBits = {
-  t_hi : bits(cap_mantissa_width - 5) = cap.T[cap_mantissa_width - 3..3];
-  t_lo : bits(3)  = cap.T[2..0];
-  b_hi : bits(cap_mantissa_width - 3) = cap.B[cap_mantissa_width - 1..3];
-  b_lo : bits(3)  = cap.B[2..0];
-  if cap.internal_e then {
-    t_lo = cap.E[5..3];
-    b_lo = cap.E[2..0];
+/* Convert from capabilty bits (128 bits with tag) to a more convenient struct. */
+function encCapabilityToCapability(t,c) : (bool, EncCapability) -> Capability = {
+  let sealed : bool = c.otype != to_bits(cap_otype_width, otype_unsealed);
+  internal_E : bool = bit_to_bool(c.internal_E);
+  E : bits(cap_E_width)  = zeros();
+  Bs : bits(cap_mantissa_width) = zeros();
+  T : bits(cap_mantissa_width - 2) = zeros();
+  lenMSBs : bits(2) = zeros();
+  if internal_E then {
+    E = c.T[internal_E_take_bits - 1..0] @ c.B[internal_E_take_bits - 1..0];
+    /* Exponent chosen so that length MSBs is 01 */
+    lenMSBs = 0b01;
+    T = c.T[cap_mantissa_width - 3..3] @ zeros(internal_E_take_bits);
+    Bs = c.B[cap_mantissa_width - 1..3] @ zeros(internal_E_take_bits);
+  } else {
+    /* Exponent zero */
+    lenMSBs = 0b00;
+    T = c.T;
+    Bs = c.B;
+  };
+  /* Reconstruct top two bits of T given T = B + len and:
+   * 1) the top two bits of B
+   * 2) most significant two bits of length derived from format above
+   * 3) carry out of B[20..0] + len[20..0] that is implied if T[20..0] < B[20..0]
+   */
+  carry_out = if T <_u Bs[cap_mantissa_width - 3..0] then 0b01 else 0b00;
```

Comment:
```suggestion
  carry_out = if T <_u Bs[cap_mantissa_width - internal_E_take_bits..0] then 0b01 else 0b00;
```

Reply:
No, I think these are unrelated. The top three bits are used to divide the representable region into eighths. While that could be parametrisable, it's independent of the number of bits used in the internal exponent I believe?

Reply:
I just saw 3 and assumed it was related.